File: proxyPattern/proxy.cpp
#include <string.h>
#include "./proxy.h"
#include "./common.h"

ImageProxy::ImageProxy(const char *fileName)
{
	// Copy fileName
	_fileName = strdup(fileName);
	_extent = Point::Zero;
	_image = 0;
}

// Control access to create object - Use cached image if exist
Image *ImageProxy::GetImage()
{
	if (_image == 0)
	{
		_image - new Image(_fileName);
	}
	return _image;
}

const Point &ImageProxy::GetExtent()
{
	// if extent is not set
	if (_extent == Point::Zero)
	{
		_extent = GetImage()->GetExtent();
	}

	// otherwise return cached extent
	return _extent;
}

void ImageProxy::Draw(const Point &at)
{
	GetImage()->Draw(at);
}

void ImageProxy::HandleMouse(Event &event)
{
	GetImage()->HandleMouse(event);
}

File: proxyPattern/csharp/Proxy.cs
// Proxy
class ImageProxy : Graphic
{
	private Image? _image;
	// Keep soft reference (not actual object) to file
	private string _fileName;
	private Point _extent;

	public ImageProxy(string fileName)
	{
		_fileName = fileName;
		_extent = new Point(0);
		// _image remain empty
	}

	// Control access to create object - Use cached image if exist
	private Image GetImage()
	{
		if (_image is null)
		{
			_image = new Image(_fileName);
		}

		return _image;
	}

	public override void Draw(Point at)
	{
		GetImage().Draw(at);
	}

	public override void Load(string from)
	{
		GetImage().Load(from);
	}

	public override void Save(string to)
	{
		GetImage().Save(to);
	}

	public override Point GetExtent()
	{
		if (_extent != null && _extent.getP() != 0)
		{
			_extent = GetImage().GetExtent();
		}

		return _extent!;
	}
}